plugin/output/postgres/postgres.go

```diff
@@ -0,0 +1,385 @@
+		outPluginType,
+		p.out,
+		nil,
+		p.controller,
+		p.config.WorkersCount_,
+		p.config.BatchSize_,
+		p.config.BatchFlushTimeout_,
+		0,
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	p.ctx = ctx
+	p.cancelFunc = cancel
+
+	p.batcher.Start(ctx)
+}
+
+func (p *Plugin) Stop() {
+	p.cancelFunc()
+	p.batcher.Stop()
+	p.pool.Close()
+}
+
+func (p *Plugin) Out(event *pipeline.Event) {
+	p.batcher.Add(event)
+}
+
+func (p *Plugin) out(_ *pipeline.WorkerData, batch *pipeline.Batch) {
+	// _ *pipeline.WorkerData - doesn't required in this plugin, we can't parse
+	// events for uniques through bytes.
+	builder := p.queryBuilder.GetInsertBuilder()
+	pgFields := p.queryBuilder.GetPgFields()
+	uniqFields := p.queryBuilder.GetUniqueFields()
+
+	// Deduplicate events, pg can't do upsert with duplication.
+	uniqueEventsMap := make(map[string]struct{}, len(batch.Events))
+	var haveValidEvents bool
+
+	for _, event := range batch.Events {
+		fieldValues, uniqueID, err := p.processEvent(event, pgFields, uniqFields)
+		if err != nil {
+			if errors.Is(err, ErrEventDoesntHaveField) {
+				stats.GetCounter(subsystemName, discardedEventCounter).Inc()
+				if p.config.Strict {
+					p.logger.Fatal(err)
+				}
+				p.logger.Error(err)
+			} else if errors.Is(err, ErrEventFieldHasWrongType) {
+				stats.GetCounter(subsystemName, discardedEventCounter).Inc()
+				if p.config.Strict {
+					p.logger.Fatal(err)
+				}
+				p.logger.Error(err)
+			}
+
+			continue
+		}
+
+		if _, ok := uniqueEventsMap[uniqueID]; ok {
+			stats.GetCounter(subsystemName, duplicatedEventCounter).Inc()
+			p.logger.Infof("event duplicated. Fields: %v, values: %v", pgFields, fieldValues)
+		} else {
+			uniqueEventsMap[uniqueID] = struct{}{}
+			haveValidEvents = true
+			builder = builder.Values(fieldValues...)
+		}
+	}
+
+	builder = builder.Suffix(p.queryBuilder.GetPostfix()).PlaceholderFormat(sq.Dollar)
+
+	// no valid events passed in non-strict mode.
+	if !haveValidEvents {
+		return
+	}
+	query, args, err := builder.ToSql()
+	if err != nil {
+		p.logger.Fatalf("Invalid SQL. query: %s, args: %v, err: %v", query, args, err)
+		return
+	}
+
+	var ctx context.Context
+	var cancel context.CancelFunc
+	// Insert into pg with retry.
+	for i := p.config.Retry; i > 0; i-- {
+		ctx, cancel = context.WithTimeout(p.ctx, p.config.DBRequestTimeout_)
+
+		p.logger.Info(query, args)
+		_, err = p.pool.ExecEx(ctx, query, &pgx.QueryExOptions{SimpleProtocol: true}, args...)
+		if err != nil {
+			cancel()
+			time.Sleep(time.Millisecond * p.config.Retention_)
+			continue
+		}
+
+		break
+	}
+	cancel()
+
+	if err != nil {
+		p.pool.Close()
+		p.logger.Fatalf("Failed insert into %s. query: %s, args: %v, err: %v", p.config.Table, query, args, err)
+	}
+}
+
+func (p *Plugin) processEvent(event *pipeline.Event, pgFields []column, uniqueFields map[string]pgType) (fieldValues []interface{}, uniqueID string, err error) {
+	fieldValues = make([]interface{}, 0, len(pgFields))
```

Comment:
to avoid allocation here we can use same logic as in 
http://github.com/ozontech/file.d/blob/master/plugin/output/kafka/kafka.go#L138-L138 and pass slice to processEvent()


smth like
```
	if *workerData == nil {
		*workerData = &data{
			fieldValues = make([]interface{}, 0, len(pgFields))
		}
	}
```

here we will need just `fieldValues = fieldValues[:0]` instead